### ISSUE TITLE
Centralized place to handle error from action

### ISSUE THREAD

**Issue by fantasia949** [developit/unistore#83]
Currently, I have to try/catch error for each action to be able to compensate or show friendly message if any error occurs.
I wonder if there is any better way to achieve that? A centralized place like store.subscribe()?

store.subscribe() is not likely for this case, it's only for state change.

**Comment by developit** [comment]
I think you could wrap `store.action`?

```js
let action = store.action;
store.action = function(...args) {
  let fn = action.apply(this, args);
  return (...args) {
    try {
      return fn.apply(this, args);
    } catch (e) {
      store.onerror();
    }
  };
}
```

**Comment by developit** [comment]
We must go deeper

**Comment by Kanaye** [comment]
hey @fantasia949!
For some reason your pen uses the old version 2.2.0.
In that [version](https://github.com/developit/unistore/blob/23b678e9a340a6f0d5fc62f69d09848ef0daf5e1/unistore.js#L26-L57) a property `action` didn't even exist on a `store`.

I updated your [pen](https://codesandbox.io/s/3k08py4y36) to the latest version and wrapping `action` works just fine in the current version.

**Comment by developit** [comment]
woot!

**Comment by fantasia949** [comment]
Thanks, @Kanaye !
I did not notice that old version. I forked that pen from repo's example, so it may need an update  😉
Your updated pen with the latest version works fine with synchronous action 👍

Unfortunately, when I continue with asynchronous version, the wrapper does not handle the rejected promise. This [pen](https://codesandbox.io/s/mq5989224j) demonstrates that.

This case is more complicated, the rejected promise cannot be handled by try/catch, and I cannot get the promise result to handle the rejection.

window.onunhandledrejection() kinda works for me, but handling from store is more preferred.

**Comment by fantasia949** [comment]
Wrapping the action inside Promise does not work.
I guess the problem is `fn.apply(this, args)` does not return any result, even with normal sync action; so the promise.catch handler cannot handle rejected promise inside the action.
You can check this [pen](https://codesandbox.io/s/ov8n2wo0ly) for more detail.

**Comment by developit** [comment]
Ah - so the above would work if we merged #54 then!

**Comment by fantasia949** [comment]
Verified the enhanced action can handle both sync and async error in version 3.0.6
